### DISCUSSION TITLE
User specific SECRET_KEY for REST API access

### DISCUSSION THREAD

**Discussion by SantoshSrinivas79** [Budibase/budibase#3368] [Help]
Hello All, I know row-level access permissions are not possible currently. Wondering if there is a way I can accomplish this.

Is there a way I can set a user-specific `SECRET_KEY`. I want to use that `SECRET_KEY` while triggering REST calls in the automation for a user.

ONLY that user with `basic` role should have access to that `SECRET_KEY`
No one else in the `basic` role should have access to it via the `api` or any other way

**Comment by mjashanks** [comment]
Hi @SantoshSrinivas79.

You could try:
1. Create a new column on the `Users` table - called `Secret`
2. Add a parameter to your RestAPI request - called `User Secret`
3. When calling your RestAPI (via `Execute Query` action), bide the value `{{ Current User.Secret }}`, to your `User Secret` parameter

**Reply by mjashanks** [reply]
P.S. My answer didn't include automations. If you need to use an automation for something, you can setup an automation with an `App Action` trigger. Then setup a parameter `User Secret` in there. The automation does not automatically know which user that executed it.

**Reply by SantoshSrinivas79** [reply]
Thanks @mjashanks .... I am currently facing major challenges with RBAC. The secrets shouldn't be accessible by other users even using the API with a `curl` command. Would really appreciate it if you could share a demo sometime.

I am absolutely not able to get the roles to work. I seem to be able to access all data with `curl` calls from the terminal.
The UI screens also do not respect the `RBAC` roles. For example, if I create a public table based on tables that require `BASIC` roles, the data is still getting shown to a `PUBLIC` user without any restrictions.

Usecase is:

- Private key `User Secret` for each user. Accessible by no one except the `Current User`
- `Current User` triggers REST calls and Automations where the REST endpoint takes the `User Secret` as part of the `Request Body`

An example would be the REST example in the docs, an individual user should be able to post issues only with their `User Secret` key. No one else should be able to impersonate them.

**Reply by shogunpurple** [reply]
Hi @SantoshSrinivas79 -

Just so I'm clear as to what you are trying to do:
- You've added a `SECRET_KEY`column to the users table in budibase
- You are firing outgoing HTTP call automations to an existing API that requires the current users `SECRET_KEY` in the request body

This part works, correct? I was just able to perform the operations that Mike detailed above.

What I'm not sure about is the use of the API with `curl` and the "secrets being accessible by other users" - do you mean that you are querying a users secret key out of budibase using our undocumented API or a webhook?

**Reply by SantoshSrinivas79** [reply]
Hi @shogunpurple, Please see below

_Yes, this is what I want to do after the RBAC works correctly (could you pls share the app if possible, will make it easier for me to try out)._

_I haven't tried out yet, since it would sense after the RBAC works as I need_

> Just so I'm clear as to what you are trying to do:
> 
> * You've added a `SECRET_KEY`column to the users table in budibase
> * You are firing outgoing HTTP call automations to an existing API that requires the current users `SECRET_KEY` in the request body
> 
> This part works, correct? I was just able to perform the operations that Mike detailed above.

_Will try out the above points after I figure out RBAC. The `SECRET_KEY` needs to be private to the logged in user._

> 
> What I'm not sure about is the use of the API with `curl` and the "secrets being accessible by other users" - do you mean that you are querying a users secret key out of budibase using our undocumented API or a webhook?

_yes correct .. I am able to query the users table using the undocumented API_ ... I seem to have access to all the data using a `BASIC` login. So I'm guessing, the `SECRET_KEY` would also be exposed.

**Reply by mjashanks** [reply]
@SantoshSrinivas79 I think you are also asking for **literally nobody** to be able to see any other user's secret - i.e. including app builders and anyone with read access to the users table.

This is not possible and the moment - anyone with read access to the Users table will see all the data in there.

Can I ask - where does this token come from? What generates it?

I'm thinking that this should be one of our considerations for our REST Api connector upgrade in December. For example, when getting a personal token using OAuth2, the user's token should always be stored completely privately. If we were to support this, it should really be its own concept, rather than a column in the user table.

**Reply by SantoshSrinivas79** [reply]
Hi @mjashanks ... you are right ... I am ok with Builders being able to access but _definitely not any user_. Is this possible quickly by any chance?

The problem I am trying to wrap my head around is that if the `Current User` needs access to the `SECRET_KEY`, based on the current permissions model will mean that **_everyone_** in the _**same role**_ as _that user_ will have access to each other's keys. Am I correct in this understanding?

For now, the `SECRET_KEY` is basically a header authorization key that my REST API endpoint will use for authorization and identification


In the long term, Your guess is exactly what I am thinking ... it is the OAuth2 token. This will make all sorts of apps be built using the current functionality of Budibase. Basically, everything that cannot be supported within Budibase can be _outsourced_ to the other app infrastructure

**Reply by mjashanks** [reply]
Hey @SantoshSrinivas79 

I have been experimenting with this just now.

In short, the secret cannot be made private at this time, sorry.

I set access to the Users table to "Power", and tried displacing the user table to a Basic user. I wasn't able to display the user data - the restriction works correctly. 

Unfortunately, I also discovered that restricting assess to the user table breaks the app in other ways. It seems like we need to be able to list all users - to be able to display user names in various places. 

It think we need a special `GET Users` endpoint, which is accessible to all logged in users, but only returns `Name` and `Id`. I think that this should be a simple enough fix, so I will log the issue and pass my thoughts on to the team.

Sorry we couldn't solve this for you now, but thanks for feeding back your use case - it's really useful stuff.

**Reply by mjashanks** [reply]
PS - I logged this issue here: https://github.com/Budibase/budibase/issues/3412

**Reply by SantoshSrinivas79** [reply]
Thank you @mjashanks @shogunpurple 

If you don't mind could you pls add a loom explanation or documentation on the `Users` table both for the app in the builder and the published app?

The table in the builder contains only the `ADMIN`, if I add other users with development access they don't show up.

I am looking for the right way to approach the `Users` table. Who creates the data in it, when it should be used? Why isn't there a "Create Row" button?

Ideally, every `User` of the app should have an entry in the `Users` table automatically right? Should that auto-creation be added?

If any, how does the data in the `Users` table change between the `dev` and `published` app?

Many thx! I can move this to a new topic if required

**Reply by SantoshSrinivas79** [reply]
Hi @mjashanks ... just checking in .... is there any support for the December release to make user `email` and other metadata incl. oauth tokens private to the logged user (in-accessible by *other* logged in users using the API calls too)

User privacy is particularly important for the use-cases I am looking for. This would be an awesome addition and I plan to handle permissions of data and everything else outside Budibase using authenticated REST apis.

Also, is there a way for passing the *current logged in* user to be sent as part of the REST call in a datasource?
(and other users should not be able to impersonate the REST call using the API or anything else)

> @SantoshSrinivas79 I think you are also asking for **literally nobody** to be able to see any other user's secret - i.e. including app builders and anyone with read access to the users table.
> 
> This is not possible and the moment - anyone with read access to the Users table will see all the data in there.
> 
> Can I ask - where does this token come from? What generates it?
> 
> I'm thinking that this should be one of our considerations for our REST Api connector upgrade in December. For example, when getting a personal token using OAuth2, the user's token should always be stored completely privately. If we were to support this, it should really be its own concept, rather than a column in the user table.
